crates/core/src/db/datastore/locking_tx_datastore/mod.rs

```diff
@@ -724,6 +694,232 @@ impl SequencesState {
     }
 }
 
+// Type aliases for lock gaurds
+type SharedReadGuard<T> = ArcRwLockReadGuard<RawRwLock, T>;
+type SharedWriteGuard<T> = ArcRwLockWriteGuard<RawRwLock, T>;
+type SharedMutexGuard<T> = ArcMutexGuard<RawMutex, T>;
+
+pub struct TxId {
+    committed_state_shared_lock: SharedReadGuard<CommittedState>,
+    lock_wait_time: Duration,
+    timer: Instant,
+}
+
+impl ReadTx for TxId {
+    fn table_id_from_name(&self, table_name: &str, database_address: Address) -> super::Result<Option<TableId>> {
+        self.committed_state_shared_lock
+            .iter_by_col_eq(
+                &ExecutionContext::internal(database_address),
+                ST_TABLES_ID,
+                NonEmpty::new(StTableFields::TableName.col_id()),
+                AlgebraicValue::String(table_name.to_owned()),
+            )
+            .map(|mut iter| {
+                iter.next()
+                    .map(|row| TableId(*row.view().elements[0].as_u32().unwrap()))
+            })
+    }
+
+    fn iter<'a>(&'a self, ctx: &'a ExecutionContext, table_id: &TableId) -> super::Result<Iter<'a>> {
+        if self.table_exists(table_id) {
+            return Ok(Iter::new(ctx, *table_id, None, &self.committed_state_shared_lock));
+        }
+        Err(TableError::IdNotFound(SystemTable::st_table, table_id.0).into())
+    }
+
+    fn table_exists(&self, table_id: &TableId) -> bool {
+        self.committed_state_shared_lock.tables.contains_key(table_id)
+    }
+
+    /// Returns an iterator,
+    /// yielding every row in the table identified by `table_id`,
+    /// where the values of `cols` are contained in `range`.
+    fn iter_by_col_range<'a, R: RangeBounds<AlgebraicValue>>(
+        &'a self,
+        ctx: &'a ExecutionContext,
+        table_id: &TableId,
+        cols: NonEmpty<ColId>,
+        range: R,
+    ) -> super::Result<IterByColRange<'a, R>> {
+        // Either the current transaction has not modified this table, or the table is not
+        // indexed.
```

Comment:
I think there is some confusion over what a `ReadTx` is

Reply:
Comment is indeed irrelevant there, will remove it but we could not have modified table because it's implemented on `TxId`, I also implemented `ReadxTx` on `MutTxId` there we can have modifed table.